atproto/firehose/client.py

```diff
@@ -217,50 +210,65 @@ def __init__(
-
-    def _on_message_callback_done(self, task: asyncio.Task) -> None:
-        self._on_message_tasks.discard(task)
+    async def _process_message_frame(self, frame: 'MessageFrame') -> None:
+        try:
+            if self._on_message_callback is not None:
+                await self._on_message_callback(frame)
+        except Exception as e:  # noqa: BLE001
+            if self._on_callback_error_callback:
+                try:
+                    await self._on_callback_error_callback(e)
+                except:  # noqa
+                    traceback.print_exc()
+            else:
+                traceback.print_exc()
 
-        exception = task.exception()
-        if exception:
-            if not self._on_callback_error_callback:
-                _print_exception(exception)
-                return
+    async def start(
+        self,
+        on_message_callback: AsyncOnMessageCallback,
+        on_callback_error_callback: t.Optional[AsyncOnCallbackErrorCallback] = None,
+    ) -> None:
+        """Subscribe to Firehose and start client.
 
-            try:
-                self._on_callback_error_callback(exception)
-            except:  # noqa
-                traceback.print_exc()
+        Args:
+            on_message_callback: Callback that will be called on the new Firehose message.
+            on_callback_error_callback: Callback that will be called if the `on_message_callback` raised an exception.
 
-    def _process_message_frame(self, frame: 'MessageFrame') -> None:
-        task: asyncio.Task = self._loop.create_task(self._on_message_callback(frame))
-        self._on_message_tasks.add(task)
-        task.add_done_callback(self._on_message_callback_done)
+        Returns:
+            :obj:`None`
+        """
+        self._on_message_callback = on_message_callback
+        self._on_callback_error_callback = on_callback_error_callback
 
-    async def start(self, *args, **kwargs):
-        super().start(*args, **kwargs)
+        self._stop_event = asyncio.Event()
 
-        while not self._stop_lock.locked():
+        while not self._stop_event.is_set():
             try:
                 if self._reconnect_no != 0:
-                    await asyncio.sleep(self._get_reconnection_delay())
+                    try:
+                        await asyncio.wait_for(self._stop_event.wait(), self._get_reconnection_delay())
+                        # Did not time out, therefore the stop event was set and we break out of the loop
+                        break
+                    except asyncio.TimeoutError:
+                        # Waited the entire delay without the stop event being set, so we continue as usual
+                        pass
 
                 async with self._get_async_client() as client:
                     self._reconnect_no = 0
 
-                    while not self._stop_lock.locked():
+                    while not self._stop_event.is_set():
                         raw_frame = await client.recv()
```

Comment:
btw probably this `.recv()` also can block for a longer time and `.stop()` will not affect immediate